File: Python/236_LowestCommonAncestorOfABinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    
    StackParents_P = list()
    StackParents_Q = list() 
    stackTransition = list()
    
    def DFS (self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode'):
        
        self.stackTransition.append(root)
        
        if (root == p):
            self.StackParents_P = list(self.stackTransition)
        elif (root == q):
            self.StackParents_Q = list(self.stackTransition)
        
        if (root.left != None):
            self.DFS(root.left, p, q)
        if (root.right != None):
            self.DFS(root.right, p, q)
        
        self.stackTransition.pop()                   
        return
        
        
    
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        # DFS keeping track of the parents.
        # Once we find a node we save the list of parents the node has.
        # Compare the lists of parents 
        self.DFS(root,p,q)      
        rtn = root
               
        
        for k in range(0, min(len(self.StackParents_P), len(self. StackParents_Q) ) ):
            if (self.StackParents_P[k] != self.StackParents_Q[k]):
                return rtn
            else:
                rtn = self.StackParents_P[k]
        return rtn

File: C#/236_LowestCommonAncestorOfABinaryTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        Stack<TreeNode> pPath = DFS(root, p);
        Stack<TreeNode> qPath = DFS(root, q);
        
        int smallerLength = pPath.Count >= qPath.Count? qPath.Count: pPath.Count;
        
        TreeNode LCA = root;
        for(int i = 0; i < smallerLength; i++) {
            TreeNode pElem = pPath.Pop();
            TreeNode qElem = qPath.Pop();
            if (pElem == qElem)
                LCA = pElem;
            else
                break;
        }
        
        return LCA;
            
    }
    
    public Stack<TreeNode> DFS(TreeNode root, TreeNode target) {
        
        if (root == target)
            return new Stack<TreeNode>(new TreeNode[] {root});
        
        if (root.left == null && root.right == null)
            return null;
        
        Stack<TreeNode> pathFound = null;
        if (root.left != null)
           pathFound = DFS(root.left, target);
        
        if (root.right != null && pathFound == null)
            pathFound = DFS(root.right, target);
        
        if (pathFound != null)
            pathFound.Push(root);
        
        return pathFound;        
    }
}